crates/ty_python_semantic/src/types/visitor.rs

```diff
@@ -198,79 +257,95 @@ fn walk_non_atomic_type<'db, V: TypeVisitor<'db> + ?Sized>(
 ///
-/// The function guards against infinite recursion
-/// by keeping track of the non-atomic types it has already seen.
-pub(super) fn any_over_type<'db>(
-    db: &'db dyn Db,
-    ty: Type<'db>,
-    query: &dyn Fn(Type<'db>) -> bool,
-) -> bool {
-    struct AnyOverTypeVisitor<'db, 'a> {
-        query: &'a dyn Fn(Type<'db>) -> bool,
+/// Note that [`TypeVisitor`] is implemented for any closure with the correct signature, meaning
+/// that you often don't need to crate a named type for your visitor.
+pub(super) fn visit_type<'db, V>(db: &'db dyn Db, ty: Type<'db>, wrapped: V)
+where
+    V: TypeVisitor<'db>,
+{
+    struct RecursionGuard<'db, V> {
+        wrapped: V,
         seen_types: FxIndexSet<NonAtomicType<'db>>,
-        found_matching_type: bool,
     }
 
-    impl<'db> TypeVisitor<'db> for AnyOverTypeVisitor<'db, '_> {
-        fn visit_type(&mut self, db: &'db dyn Db, ty: Type<'db>) {
-            if self.found_matching_type {
-                return;
-            }
-            self.found_matching_type |= (self.query)(ty);
-            if self.found_matching_type {
-                return;
-            }
+    impl<'db, V> TypeVisitor<'db> for RecursionGuard<'db, V>
+    where
+        V: TypeVisitor<'db>,
+    {
+        fn visit_type(&mut self, db: &'db dyn Db, ty: Type<'db>) -> TypeVisitorResult {
+            self.wrapped.visit_type(db, ty)?;
             match TypeKind::from(ty) {
                 TypeKind::Atomic => {}
                 TypeKind::NonAtomic(non_atomic_type) => {
-                    if !self.seen_types.insert(non_atomic_type) {
-                        // If we have already seen this type, we can skip it.
-                        return;
+                    if self.seen_types.insert(non_atomic_type) {
+                        // If we haven't already seen this type, we should recurse into it.
+                        walk_non_atomic_type(db, non_atomic_type, self)?;
                     }
-                    walk_non_atomic_type(db, non_atomic_type, self);
                 }
             }
+            Ok(())
         }
     }
 
-    let mut visitor = AnyOverTypeVisitor {
-        query,
+    let mut visitor = RecursionGuard {
+        wrapped,
         seen_types: FxIndexSet::default(),
-        found_matching_type: false,
     };
-    visitor.visit_type(db, ty);
-    visitor.found_matching_type
+    let _ = visitor.visit_type(db, ty);
+}
+
+/// Return `true` if `ty`, or any of the types contained in `ty`, match the closure passed in.
+pub(super) fn any_over_type<'db>(
+    db: &'db dyn Db,
+    ty: Type<'db>,
+    query: &'db dyn Fn(Type<'db>) -> bool,
+) -> bool {
+    let mut found_matching_type = false;
+    visit_type(db, ty, |_, ty| {
+        if found_matching_type {
+            return Err(AbortTypeVisitor);
+        }
+        found_matching_type |= query(ty);
+        if found_matching_type {
+            return Err(AbortTypeVisitor);
+        }
+        Ok(())
+    });
```

Comment:
Here is what it looks like to use the new recursion guard helper with a closure that automatically implements the `TypeVisitor` trait.